src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSpark.java

```diff
@@ -80,18 +80,25 @@ protected void runTool(final JavaSparkContext ctx) {
         final OpticalDuplicateFinder finder = opticalDuplicatesArgumentCollection.READ_NAME_REGEX != null ?
                 new OpticalDuplicateFinder(opticalDuplicatesArgumentCollection.READ_NAME_REGEX, opticalDuplicatesArgumentCollection.OPTICAL_DUPLICATE_PIXEL_DISTANCE, null) : null;
 
-        final JavaRDD<GATKRead> finalReads = mark(reads, getHeaderForReads(), duplicatesScoringStrategy, finder, parallelism);
-
-        try {
-            ReadsSparkSink.writeReads(ctx, output, finalReads, getHeaderForReads(), parallelism == 1 ? ReadsWriteFormat.SINGLE : ReadsWriteFormat.SHARDED);
-        } catch (IOException e) {
-            throw new GATKException("unable to write bam: " + e);
-        }
+        final JavaRDD<GATKRead> finalReadsForMetrics = mark(reads, getHeaderForReads(), duplicatesScoringStrategy, finder, parallelism);
 
         if (metricsFile != null) {
-            final JavaPairRDD<String, DuplicationMetrics> metricsByLibrary = MarkDuplicatesSparkUtils.generateMetrics(getHeaderForReads(), finalReads);
+            final JavaPairRDD<String, DuplicationMetrics> metricsByLibrary = MarkDuplicatesSparkUtils.generateMetrics(getHeaderForReads(), finalReadsForMetrics);
             final MetricsFile<DuplicationMetrics, Double> resultMetrics = getMetricsFile();
             MarkDuplicatesSparkUtils.saveMetricsRDD(resultMetrics, getHeaderForReads(), metricsByLibrary, metricsFile, getAuthHolder());
         }
+
+        //The OD attribute was added to each read for optical dups.
+        // Now we have to clear it to avoid polluting the output.
+        final JavaRDD<GATKRead> finalReads = finalReadsForMetrics.map(read -> {
+            read.clearAttribute(MarkDuplicatesSparkUtils.OPTICAL_DUPLICATE_TOTAL_ATTRIBUTE_NAME);
+            return read;
+        });
```

Comment:
yes, we add it temporarily during computation. Non-spark does not do it.
It's polluting because each reads is printed out with the number of reads that are its optical dups, which is 0 for almost every read. It makes the output different than non-spark